wallet: birth time update during tx scanning

furszy opened the pull request:
Fixing #28897.

As the user may have imported a descriptor with a timestamp newer
than the actual birth time of the first key (by setting 'timestamp=now'),
the wallet needs to update the birth time when it detects a transaction
older than the oldest descriptor timestamp.

Testing Notes:
Can cherry-pick the test commit on top of master. It will fail there.

furszy committed:
test: coverage for wallet birth time interaction with -reindex

Verifying the wallet updates the birth time accordingly when it
detects a transaction with a time older than the oldest descriptor
timestamp.
This could happen when the user blindly imports a descriptor with
'timestamp=now'.

furszy committed:
wallet: birth time update during tx scanning

As the user could have imported a descriptor with
a newer timestamp (by blindly setting 'timestamp=now'),
the wallet needs to update the birth time when it detects
a transaction older than the oldest descriptor timestamp.

furszy committed:
refactor: rename FirstKeyTimeChanged to MaybeUpdateBirthTime

In the following-up commit, the wallet birth time will also
be modified by the transactions scanning process. When a tx
older than all descriptor's timestamp is detected.

Sjors commented:
Concept ACK

You could also add a `oldest_timestamp` (or `birth`) field to `getwalletinfo` which would contain the birth date.

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -3516,7 +3516,7 @@ void CWallet::ConnectScriptPubKeyManNotifiers()
         spk_man->NotifyWatchonlyChanged.connect(NotifyWatchonlyChanged);
         spk_man->NotifyCanGetAddressesChanged.connect(NotifyCanGetAddressesChanged);
-        spk_man->NotifyFirstKeyTimeChanged.connect(std::bind(&CWallet::FirstKeyTimeChanged, this, std::placeholders::_1, std::placeholders::_2));
+        spk_man->NotifyFirstKeyTimeChanged.connect(std::bind(&CWallet::TryUpdateBirthTime, this, std::placeholders::_2));
```
4bfb243ac41cf1396cf55f8b24de2ba52186c7a4: you can use `std::placeholders::_1` here, by dropping `this` from `NotifyFirstKeyTimeChanged` calls and changing:

```diff
--- a/src/wallet/scriptpubkeyman.h
+++ b/src/wallet/scriptpubkeyman.h
@@ -255,11 +255,11 @@ public:
 
     /** Keypool has new keys */
     boost::signals2::signal<void ()> NotifyCanGetAddressesChanged;
 
     /** Birth time changed */
-    boost::signals2::signal<void (const ScriptPubKeyMan* spkm, int64_t new_birth_time)> NotifyFirstKeyTimeChanged;
+    boost::signals2::signal<void (int64_t new_birth_time)> NotifyFirstKeyTimeChanged;
 };

```

Sjors replied:
If you have good reasons to believe we need these events later, then keeping them is fine by me. But why do you think this?

furszy replied:
Yeah, I was thinking about https://github.com/bitcoin/bitcoin/pull/28920#discussion_r1401323753 and other features I have in mind. But, more generally, I believe the wallet should be aware of the source of the event. However, upon further thought, we could achieve this by creating multiple handlers on the wallet side instead of passing the origin pointer to the signal itself.

Sjors reviewed (commented):
I checked that the test fails if you drop everything but 134960c3828abe658d565af52e3c6c5af03019f2.

furszy committed:
rpc: getwalletinfo, return wallet 'birthtime'

And add coverage for it

furszy commented on src/wallet/wallet.h:
```diff
@@ -686,7 +686,7 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
 
     /** Updates wallet birth time if 'new_birth_time' is below it */
-    void FirstKeyTimeChanged(const ScriptPubKeyMan* spkm, int64_t new_birth_time);
+    void TryUpdateBirthTime(int64_t new_birth_time);
```
sure, updated.

furszy commented:
Updated per feedback, thanks @Sjors.

> You could also add a oldest_timestamp (or birth) field to getwalletinfo which would contain the birth date.

Sounds good 👌🏼.
Added it in another commit to not disable the possibility of verifying the failure by cherry-picking the test commit on top of master.

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -3483,7 +3483,7 @@ void CWallet::AddScriptPubKeyMan(const uint256& id, std::unique_ptr<ScriptPubKey
     const auto& spkm = m_spk_managers[id] = std::move(spkm_man);
```
Ah, that makes sense, can you add a comment? Something like:
```cpp
// Add spkm_man to m_spk_managers before calling any method
// that might access it.
const auto& spkm = m_spk_managers[id] = std::move(spkm_man);
```

furszy commented:
Updated per feedback. Tiny [diff](https://github.com/bitcoin/bitcoin/compare/44e2c744b8bae075da69d8e8dd392296899a68a5..25de6ff824bffd8617a6baca19bbe57d6764a9d2).
Only added a small explanatory comment.

furszy commented on test/functional/wallet_reindex.py:
```diff
@@ -0,0 +1,93 @@
+        # Fund address to test
+        wallet_addr = miner_wallet.getnewaddress()
+        tx_id = miner_wallet.sendtoaddress(wallet_addr, 2)
+        miner_wallet.syncwithvalidationinterfacequeue()
```
> q: Is this needed? It should be called internally for most stuff that needs it.

Good eye. No, it is not needed.
I was probably worried about the tx not reaching the mempool prior to the blocks generation when I did it but.. `sendtoaddress` (same as every other send mechanism) always reaches the mempool. Returning a mempool error if it fails to get added there.
And.. this is independent to the validation interface queue. So, I'm not sure why my past-self added it.

furszy replied:
Pushed it. Thanks.

BrandonOdiwuor reviewed (commented):
Concept ACK

Sjors commented:
ACK 8cd2fedb3ab283f9dbabd8cf1827e63f4f0db8d7

BrandonOdiwuor commented on test/functional/wallet_reindex.py:
```diff
@@ -0,0 +1,92 @@
+            self.advance_time(node, BLOCK_TIME)
+
+        # Now create a new wallet, and import the descriptor
+        node.createwallet(wallet_name='watch_only', disable_private_keys=True, blank=True, load_on_startup=True)
```
```suggestion
        node.createwallet(wallet_name='watch_only', disable_private_keys=True, load_on_startup=True)
```

nit: blank is not necessary.

furszy replied:
Done

BrandonOdiwuor reviewed (approved):
ACK 8cd2fedb3ab283f9dbabd8cf1827e63f4f0db8d7

left a nit comment

achow101 commented on test/functional/test_runner.py:
```diff
@@ -206,6 +206,7 @@
     'wallet_createwallet.py --descriptors',
     'wallet_watchonly.py --legacy-wallet',
     'wallet_watchonly.py --usecli --legacy-wallet',
+    'wallet_reindex.py',
```
In bc8fb2a9d5fe941bcc77eacb23d72c663828f8e3 "test: coverage for wallet birth time interaction with -reindex"

Needs `--legacy-wallet` and `--descriptor`.

furszy replied:
Done.
Plus, discovered and fixed an issue with the legacy wallet birth time along the way.

furszy committed:
wallet: fix legacy spkm default birth time

To avoid scanning blocks, as assumed by a wallet with no
generated keys or imported scripts, the default value for
the birth time needs to be set to the maximum int64_t value.

Once the first key is generated or the first script is imported,
the legacy SPKM will update the birth time automatically.

furszy commented:
Updated per feedback, thanks for the reviews.

While updating the test suite to incorporate the `--legacy-wallet` test counterpart, discovered an issue with the first key time in blank legacy wallets. It was incorrectly set to 0 (aka scan from the beginning of time) instead of being set to the maximum timestamp value. Therefore, in master, blank legacy wallets are currently scanning the entire chain instead of utilizing the birth time functionality. Added 0a1107d to fix the issue.

luke-jr commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -132,6 +133,9 @@ static RPCHelpMan getwalletinfo()
     obj.pushKV("descriptors", pwallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS));
     obj.pushKV("external_signer", pwallet->IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER));
     obj.pushKV("blank", pwallet->IsWalletFlagSet(WALLET_FLAG_BLANK_WALLET));
+    if (int64_t birthtime = pwallet->GetBirthTime(); birthtime != std::numeric_limits<int64_t>::max()) {
```
Shouldn't this be `NO_KEY_TIME` explicitly?

furszy replied:
yeah, it could be `NO_KEY_TIME` or could also return an optional.

The issue that I see with the `NO_KEY_TIME` constant is where to place it. Because, if we place it at the wallet.h level, we would add a circular dependency wallet -> spkm -> wallet.

achow101 replied:
Agree that this should be `NO_KEY_TIME`. Since wallet already includes spkm, leaving it there will not be circular, and it can still be accessed by things that include wallet.

furszy replied:
Done. Updated.

luke-jr commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -286,7 +286,9 @@ class LegacyScriptPubKeyMan : public ScriptPubKeyMan, public FillableSigningProv
 
-    int64_t nTimeFirstKey GUARDED_BY(cs_KeyStore) = 0;
+    // By default, do not scan any block until keys/scripts are generated/imported
+    int64_t NO_KEY_TIME = std::numeric_limits<int64_t>::max();
```
Seems like `NO_KEY_TIME` ought to be `0` so it triggers a full scan

furszy replied:
> Seems like `NO_KEY_TIME` ought to be `0` so it triggers a full scan

That was the bug I solved 0a1107d0.
By default, the legacy spkm is empty. It has no keys nor scripts, thus it should not trigger a full scan.
Once the first key is generated or the first script is imported,
the legacy spkm will update the `nTimeFirstKey` time automatically and start scanning blocks.

S3RK commented:
Approach ACK

Sjors commented:
re-utACK 1ce45baed7dd2da3f1cb85c9c25110e5537451ae

achow101 commented:
> Should we backport [6f49737](https://github.com/bitcoin/bitcoin/commit/6f497377aa17cb8a590fd7717fa8ededf4249999) or the whole PR, and if so how far?

I think it would be good to backport the entire thing.

achow101 commented:
ACK 1ce45baed7dd2da3f1cb85c9c25110e5537451ae

fanquake commented:
Adding to #29011 for backporting to 26.x.
